Revised partitioning algorithm

amaleewilson opened the pull request:
This adds the "revised" partitioning algorithm to the splitter and adds support for collecting literals from the sat solver. While there is only one partitioning strategy in this PR, I have like half a dozen, and I'm trying to make them easier to read. 

Something I'm wondering about is whether nodebuilder or nodemanager->mkNode is a better strategy for node creation. A nice thing about mkAnd, for example, is that I don't have to explicitly handle the case where only 2 partitions are requested because mkAnd has the appropriate behavior when there is only one node. I think we used nodebuilder in the past because it was supposed to be faster, but I'm not sure.

amaleewilson committed:
first version before PR

amaleewilson committed:
first version before PR

amaleewilson committed:
first version before PR

amaleewilson committed:
first version before PR

ajreynol committed:
[proofs] Consolidate multiple substitutions to single AND proof in trust substitution map (#8109)

This ensures that a common proof node is used for a substitution instead of copying its children. This optimization had been noted in a comment but had not yet been enabled.

This should speed up proof reconstruction for substitutions.

Lachnitt committed:
[proofs] [alethe] Introduce all_simplify and replace old trust rules (#8100)

Instead of printing "trust" in the case of a simplification rule that cannot be translated to an Alethe simplify rule print "all_simplify". This also replaces the heuristic in THEORY_REWRITE to determine which simplify rule is needed.

ajreynol committed:
Remove some irrelevant node kinds from the model (#8110)

There are several kinds of nodes that are currently sent to the model that don't need to be. Furthermore, we require not sending some of these due to the new invariants in #8095.

ajreynol committed:
Missing ids for arith conflicts (#8108)

Fixes #8097.

ajreynol committed:
Introduce skolem function to make transcendental function purification consistent (#8116)

Previously, we could introduce multiple purifications for the same sin term.

4tXJ7f committed:
Fix `STRINGS_ARRAY_NTH_UPDATE` inference (#8121)

Our `STRINGS_ARRAY_NTH_UPDATE` inference was missing an explanation that
establishes the relation between the update and the nth term. On the
current set of benchmarks, this does not cause an issue, but there
should be cases where it does. The commit also changes the array core solver
to take better advantage of `d_indexMap`.

ajreynol committed:
Make spurious assertion into warning (#8051)

Fixes cvc5/cvc5-projects#401.

4tXJ7f committed:
Improve `STRINGS_ARRAY_UPDATE_BOUND` inference (#8123)

This changes the `STRINGS_ARRAY_UPDATE_BOUND` inference to explicitly
include the disequality between the term before and after the update if
the update has an effect. This is an optimization that prevents us from
splitting on that (dis-)equality later on.

ajreynol committed:
Add well formed term check to solver engine (#8056)

Fixes cvc5/cvc5-projects#413.

ajreynol committed:
Throw option exceptions when combining input conversion preprocessing with sygus (#8059)

Fixes cvc5/cvc5-projects#436.

ajreynol committed:
Add unit test for fixed issue with get-difficulty (#8060)

Fixed by recent updates to get-difficulty.

Fixes cvc5/cvc5-projects#434.

ajreynol committed:
Fixes and additions for LFSC signatures (#8120)

Fixes a critical issue in the re intersection rule, where a null terminator was missed.

Also adds a sketch of the theory signature for floating points.

vinciusb committed:
[proofs] [dot] Enable DAG and tree printing with dot files (#8107)

Add a proof option that allows DAG and tree print format when using dot files.

Signed-off-by: Vinícius Braga Freire vinicius.braga.freire@gmail.com

ajreynol committed:
Relax what is generated in the model from NL (#8113)

This makes it so that we don't assign concrete values to equivalence classes with transcendental functions.

It also removes some unused infrastructure.

amaleewilson committed:
Fix warnings, clean up decision collector

amaleewilson committed:
Fix warnings, clean up decision collector

amaleewilson committed:
format

amaleewilson committed:
format

amaleewilson committed:
format

amaleewilson committed:
format

ajreynol committed:
Change inference scheme in transcendentals to rewrite rule (#8115)

Avoids introducing new sine terms.

ajreynol committed:
Remove refineConflicts option (#8129)

Was introduced in 8a0c056.

Fixes #8119.

amaleewilson committed:
refactor and format

amaleewilson committed:
refactor and format

amaleewilson committed:
fixed case for 2 partitions

amaleewilson committed:
fixed case for 2 partitions

amaleewilson committed:
remove old commented out experiment

amaleewilson committed:
remove old commented out experiment

amaleewilson committed:
remove old commented out experiment

amaleewilson committed:
remove old commented out experiment

amaleewilson committed:
remove old commented out experiment

amaleewilson committed:
remove old commented out experiment

amaleewilson committed:
a few modifications

amaleewilson committed:
a few modifications

ajreynol committed:
Support some cases of isConst for regular expressions (#8114)

Fixes https://github.com/cvc5/cvc5/issues/8111.

Having `isConst` return true is required for some rare use cases of regular expressions, e.g. rewrite rule synthesis involving regular expression variables.

amaleewilson committed:
comments

amaleewilson committed:
comments

amaleewilson committed:
comment

amaleewilson committed:
comment

amaleewilson commented on src/prop/theory_proxy.cpp:
```diff
@@ -52,11 +52,11 @@ TheoryProxy::TheoryProxy(Env& env,
+  // bool trackTopLevelLearned = isOutputOn(OutputTag::LEARNED_LITS);
+  // if (trackTopLevelLearned)
+  // {
     d_zll = std::make_unique<ZeroLevelLearner>(env, propEngine);
```
Not relevant for this particular partitioning algorithm, but I want to always collect the learned literals for a different partitioning strategy.

4tXJ7f replied:
Hmm, we shouldn't comment out code in a PR, can we just change the condition of the `if` statement to also include certain partitioning strategies?

amaleewilson replied:
Good idea. I also might have a totally separate flag for appending the learned literals, but I will put this back for now until I add those partitioning strategies / different flags.

amaleewilson committed:
Fixed a refactoring bug

amaleewilson committed:
Fixed a refactoring bug

4tXJ7f commented on src/theory/theory_engine.h:
```diff
@@ -403,6 +404,10 @@ class TheoryEngine : protected EnvObj
+
+  std::unique_ptr<cvc5::theory::Splitter> d_splitter;
+  int d_numPartitions;
+  std::list<Node> d_assertedPartitions;
```
Data members [go last](https://github.com/cvc5/cvc5/wiki/Code-Style-Guidelines#declaration-order). It would also be good to add a short comment to those.

4tXJ7f commented on src/theory/theory_engine.cpp:
```diff
@@ -392,8 +400,22 @@ void TheoryEngine::check(Theory::Effort effort) {
+      }
     }
 
+    if (options::computePartitions() > 1 && options::partitionCheck() == "standard"){
```
Apply ClangFormat.

4tXJ7f commented on src/options/smt_options.toml:
```diff
@@ -9,6 +9,58 @@ name   = "SMT Layer"
+  name       = "partitionStrategy"
+  category   = "expert"
+  long       = "partition-strategy=X"
+  type       = "std::string"
```
Instead of using an `std::string`, it would be better to use a "mode" option. See [here](https://github.com/cvc5/cvc5/blob/master/src/options/decision_options.toml#L4-L27) for an example.

4tXJ7f commented on src/prop/prop_engine.h:
```diff
@@ -142,6 +142,8 @@ class PropEngine : protected EnvObj
    */
   bool isDecision(Node lit) const;
 
+  std::vector<Node> getPropDecisions() const;
```
Add a comment for this method.

4tXJ7f commented on src/prop/sat_solver.h:
```diff
@@ -143,6 +143,8 @@ class CDCLTSatSolverInterface : public SatSolver
 
   virtual bool isDecision(SatVariable decn) const = 0;
 
+  virtual std::vector<SatLiteral> getDecisions() const = 0;
```
Add a comment for this method.

4tXJ7f commented on src/prop/minisat/minisat.cpp:
```diff
@@ -266,6 +266,17 @@ bool MinisatSatSolver::isDecision(SatVariable decn) const {
+{
+  std::vector<SatLiteral> decisions;
+  Minisat::vec<Minisat::Lit> miniDecisions;
+  d_minisat->getMiniSatDecisions().copyTo(miniDecisions);
```
Is this copy necessary?

amaleewilson replied:
nope

4tXJ7f commented on src/theory/splitter.h:
```diff
@@ -0,0 +1,93 @@
+        d_numChecks(0),
+        d_numPartitionsSoFar(0),
+        d_partitionFile(options::writePartitionsToFileName())
+  {
```
We tend to have all the function definitions in the source file for anything that is longer than a single line.

4tXJ7f commented on src/theory/splitter.h:
```diff
@@ -0,0 +1,93 @@
+    }
+  }
+
+  TrustNode makePartitions();
```
Add a comment.

4tXJ7f commented on src/theory/splitter.h:
```diff
@@ -0,0 +1,93 @@
+  TrustNode makePartitions();
+
+ private:
+  prop::PropEngine* d_propEngine;
+  std::unique_ptr<Valuation> d_valuation;
+  const unsigned d_numPartitions;
+  unsigned d_numChecks;
+  unsigned d_numPartitionsSoFar;
+  std::string d_partitionFile;
+  std::ofstream d_partitionFileStream;
+  std::ostream* d_output;
+  std::list<Node> d_assertedLemmas;
+  std::vector<Node> d_cubes;
+  unsigned d_conflictSize;
+  void emitCube(Node toEmit);
+  TrustNode blockPath(Node toBlock);
+  void closeFile();
+  TrustNode stopPartitioning();
+  TrustNode handlePartitionTwoOfTwo();
+  void collectDecisionLiterals(std::vector<TNode>& literals);
```
Add comments and [move the methods before the data members](https://github.com/cvc5/cvc5/wiki/Code-Style-Guidelines#declaration-order).

4tXJ7f commented on src/theory/theory_engine.cpp:
```diff
@@ -200,6 +201,13 @@ void TheoryEngine::finishInit()
   }
+
+  // Won't work with incremental solving, but the partitioning
+  // is focused on non-incremental anyway. 
```
It would be good to add an options check to [this](https://github.com/cvc5/cvc5/blob/33fab4e47b9f5cc7dae76a30d7850a82bf8290d7/src/smt/set_defaults.cpp#L976) method.

4tXJ7f commented on src/theory/splitter.cpp:
```diff
@@ -0,0 +1,184 @@
+      // of the requested number of partitions.
+      if (literals.size() < d_conflictSize) return TrustNode::null();
+
+      std::vector<Node> tmpLiterals(literals.begin(),
+                                    literals.begin() + d_conflictSize);
```
Instead of copying the literals, why not resize the vector?

4tXJ7f commented on src/theory/splitter.cpp:
```diff
@@ -0,0 +1,184 @@
+// Emit the negation of all previously asserted lemmas.
+TrustNode Splitter::stopPartitioning()
+{
+  NodeBuilder andBuilder(kind::AND);
+  for (const auto d : d_assertedLemmas) andBuilder << d;
+  Node conj = andBuilder.constructNode();
+  NodeBuilder notBuilder(kind::NOT);
+  notBuilder << conj;
+  Node lemma = notBuilder.constructNode();
```
```suggestion
  Node lemma = NodeManager::currentNM()->mkAnd(d_assertedLemmas).notNode();
```

4tXJ7f reviewed (changes requested):
Just some quick superficial comments.

4tXJ7f commented:
> Something I'm wondering about is whether nodebuilder or nodemanager->mkNode is a better strategy for node creation. A nice thing about mkAnd, for example, is that I don't have to explicitly handle the case where only 2 partitions are requested because mkAnd has the appropriate behavior when there is only one node. I think we used nodebuilder in the past because it was supposed to be faster, but I'm not sure.

`NodeManager::mkNode()` is internally using a `NodeBuilder`, so there shouldn't really be any performance difference for regular cases. `NodeBuilder` can be faster if you first create an `std::vector` of `Node`s that you pass to `NodeManager::mkNode()`, but I wouldn't worry about it for now.

ajreynol committed:
Fix issue in datatypes care graph computation involving subtyping (#8125)

The datatypes care graph was incorrectly computed for constructors taking `Real` arguments, due to subtyping.

This fixes the term index to cache on the *return* type for constructors, not the argument types.  Caching based on argument types was incorrect even for non-parametric datatypes with arguments that have subtyping.

Fixes https://github.com/cvc5/cvc5/issues/8124.

This also does minor cleanup and additions done while debugging the issue.

nafur committed:
Refactor multiplication in arithmetic rewriter (#7965)

This PR refactors the rewriting of multiplication. Most importantly, we explicitly deal with distributivity of addition and multiplication explicitly using the new utilities.

nafur committed:
Fix pruning of covering intervals in proofs (#8084)

We need to mirror the pruning of intervals in the coverings solver in the corresponding proof. To match the proof nodes with the coverings-internal intervals, we use ids, where non-interval nodes have the id zero. In this process, we had three separate issues that never showed up because we don't have a proof checker for CAD proofs:

we did not check for zeros, pruning all non-interval nodes
we ran the pruning on the direct children of a CAD_RECURSIVE proof step, which are SCOPE nodes that don't have ids
iterating over and then removing children was done by a custom but faulty reimplementation of std::remove_if
Thus, we now

never prune zero nodes,
prune based on the (single) child node, if we are checking a SCOPE node, and
simply use std::remove_if.

4tXJ7f committed:
[Rewriter] Do not attempt to rewrite constants (#8061)

This commit changes our rewriter policy to not attempt to rewrite nodes
with no children.

ajreynol committed:
Properly sanatize user names in LFSC (#8080)

Previously, we would get LFSC proof checking failures if we were using an identifier that was valid in SMT-LIB but not LFSC.

Co-authored-by: Alex Ozdemir <aozdemir@hmc.edu>

nafur committed:
Fix icp candidate parsing (#8137)

For constraints of the form c * x ~ poly(x) where c != 1 we would sometimes forget to take the inverse of c when parsing the assertion to the internal representation of the ICP solver. This PR fixes this issue, and does a few more changes:

--nl-icp is really not well-tested and disabled by default. We make it an expert option.
we no longer issue the propagation lemmas if we already found a conflict
the nl-icp trace was heavily polluted by output from the interval utils, which are now moved to a new nl-icp-debug trace
Fixes #8135.

nafur committed:
Fix creation of RAN from non-dyadic rational (#8138)

Libpoly only allows to create real algebraic numbers from dyadic rationals. For non-dyadic rationals, we need to create a polynomials and create it the general way. Libpoly requires all defining polynomials to be primitive, in particular the leading coefficient must be positive. We would always have it negative, which only becomes a problem if libpoly has assertions enabled, though.

Fixes #8126.

ajreynol committed:
Option exception when incompatible with proofs (#8064)

This changes set defaults so that it doesn't silently disable proofs or unsat cores.

Fixes cvc5/cvc5-projects#440.

nafur committed:
Remove long obsolete unsafe interrupt exception (#8139)

We used to use UnsafeInterruptException to deal with resource outs in the resource manager: when resources were exhausted, we would throw this exception that was catched in the core solver engine. This had the significant downside of leaving the solver in a potentially inconsistent state. We moved away from using it long ago in #4732.
What remained was the UnsafeInterruptException class itself and a bunch of places that still catch this exception that is no longer thrown anywhere.

This PR removes this class for good.

ajreynol committed:
Eliminate match from LFSC proofs (#8090)

The smt 2.6 term match is very hard to represent in LFSC, this eliminates it in favor of the ITE term that it is syntax sugar for. Like other aspects of the LFSC conversion, this is part of the trusted core.

This avoids internal type errors in the LFSC node converter.

ajreynol committed:
Do not insist that entries for UF are constant in FMF (#8140)

Fixes #8096.

ajreynol committed:
Further relax what is considered a value in the model (#8095)

Fixes #8094.

This makes it so that "value-like" terms can appear as subterms in other terms that are then subsequently also considered values. For example (str.++ (witness ((x String)) (= (str.len x) 1000)) "A") is now considered a value.

This also changes the default option for when to use witness terms for strings in models, based on some user feedback.

amaleewilson committed:
Clean up

amaleewilson committed:
Clean up

amaleewilson committed:
Addressed all comments

amaleewilson committed:
Addressed all comments

amaleewilson committed:
Merge branch 'partitioning' of github.com:amaleewilson/cvc5 into partitioning

amaleewilson committed:
Revert "Merge branch 'partitioning' of github.com:amaleewilson/cvc5 into partitioning"

This reverts commit e29d672ac5f3b5cb9756d87a42446220ad1237ce, reversing
changes made to 13c9bf3f593fd734d6ae996ae098d25177683a19.

amaleewilson commented:
I messed up rebasing, so I'm closing this PR And will open another one.
